### ISSUE TITLE
NODEFS: Unable to create file with O_TRUNC if filename is the previous name of a renamed file

### ISSUE THREAD

**Issue by petersalomonsen** [emscripten-core/emscripten#8187]
Opening a non-existing file with `O_CREAT|O_TRUNC` does and should work. However if the filename is the same as another file had before rename it fails with NODEFS.

Here's how to reproduce:

```
#include "emscripten.h"
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>

int main() {
    printf("Make sure you have removed test.txt and renamedtest.txt before running.\n");
    EM_ASM_({
		FS.mkdir("/work");
		FS.mount(NODEFS, {root: '.'},"/work");
		FS.chdir("/work");
    });

    int fd = open("test.txt",
			 O_CREAT|O_TRUNC|O_WRONLY,
			 0666);
    printf("Opening a non-existing file with O_TRUNC is ok: %d\n", fd);
    
    write(fd, "test", 4);
    close(fd);
    
    // Renaming the file will cause opening the original filename with O_TRUNC to fail
    rename("test.txt", "renamedtest.txt");
    
    fd = open("test.txt",
			 O_CREAT|O_TRUNC|O_WRONLY,
			 0666);

    // FD is -1 which it should not
    printf("FD is negative when opening a non-existing file with a filename that existed previously with O_TRUNC: %d\n", fd);

    fd = open("test.txt",
			 O_CREAT|O_WRONLY,
			 0666);

    printf("FD is OK if opening the same file without O_TRUNC: %d\n", fd);
}
```

The bug was identified when compiling [Xapian](https://github.com/xapian/xapian) with emscripten and a workaround is currently applied for this problem: https://github.com/xapian/xapian/blob/master/xapian-core/backends/glass/glass_version.cc#L298

**Comment by ojwb** [comment]
I've now updated the comment for the workaround in the Xapian code.

It occured to that the [handling for `O_TRUNC` here](https://github.com/emscripten-core/emscripten/blob/incoming/src/library_fs.js#L1049) could probably be moved inside the `if (!created) {`...`}` check that's right before it, since if we just created the file then it'll be zero size so there's no need to truncate it.

That would save a little unnecessary work and presumably stop this manifesting in our case, but it seems the deeper problem is that something remains attached to the old name when a file is renamed, which is probably going to cause problems in other cases if the filename is later used again.
